File: WPF/Controls/EqualizingStackPanel.vb
Namespace Controls

    Public Class EqualizingStackPanel
        Inherits Panel

        ' As Orientation.Horizontal is 0 (which makes no changes to the Orientation when got XOred with it), then the default for writing the Measure and Arrange methods will be the horizontal orientation.

        Protected Overrides Function MeasureOverride(ByVal AvailableSize As Size) As Size
            Dim ChildCount = Me.Children.Count

            If ChildCount = 0 Then
                Return Size.Empty
            End If

            Dim MaxHeight = 0.0
            Dim MaxWidth = 0.0
            AvailableSize = Me.CreateSize(Me.Dimension(AvailableSize, Orientation.Horizontal) / ChildCount,
                                          Me.Dimension(AvailableSize, Orientation.Vertical))

            For Each C As UIElement In Me.Children
                C.Measure(AvailableSize)
                Dim Sz = C.DesiredSize
                MaxWidth = Math.Max(MaxWidth, Me.Dimension(Sz, Orientation.Horizontal))
                MaxHeight = Math.Max(MaxHeight, Me.Dimension(Sz, Orientation.Vertical))
            Next

            Return Me.CreateSize(MaxWidth * ChildCount, MaxHeight)
        End Function

        Protected Overrides Function ArrangeOverride(ByVal FinalSize As Size) As Size
            Dim ChildCount = Me.Children.Count
            Dim OrigFinalSize = FinalSize

            Dim Width = Me.Dimension(FinalSize, Orientation.Horizontal) / ChildCount
            FinalSize = Me.CreateSize(Width, Me.Dimension(FinalSize, Orientation.Vertical))
            Dim X = 0.0

            For Each C As UIElement In Me.Children
                C.Arrange(New Rect(Me.CreatePoint(X, 0), FinalSize))
                X += Width
            Next

            Return OrigFinalSize
        End Function

        Private Function CreateSize(ByVal Width As Double, ByVal Height As Double) As Size
            If Me.OrientationCache = Orientation.Horizontal Then
                Return New Size(Width, Height)
            Else
                Return New Size(Height, Width)
            End If
        End Function

        Private Function CreatePoint(ByVal X As Double, ByVal Y As Double) As Point
            If Me.OrientationCache = Orientation.Horizontal Then
                Return New Point(X, Y)
            Else
                Return New Point(Y, X)
            End If
        End Function

        Private ReadOnly Property Dimension(ByVal Size As Size, ByVal Orientation As Orientation) As Double
            Get
                Orientation = Orientation Xor Me.OrientationCache
                If Orientation = Orientation.Horizontal Then
                    Return Size.Width
                Else
                    Return Size.Height
                End If
            End Get
        End Property

#Region "Orientation Property"
        Public Shared ReadOnly OrientationProperty As DependencyProperty = DependencyProperty.Register("Orientation", GetType(Orientation), GetType(EqualizingStackPanel), New PropertyMetadata(Orientation.Vertical, AddressOf Orientation_Changed, AddressOf Orientation_Coerce))

        Private Shared Function Orientation_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            'Dim Self = DirectCast(D, EqualizingStackPanel)

            'Dim Value = DirectCast(BaseValue, Orientation)

            Return BaseValue
        End Function

        Private Shared Sub Orientation_Changed(ByVal D As DependencyObject, ByVal E As DependencyPropertyChangedEventArgs)
            Dim Self = DirectCast(D, EqualizingStackPanel)

            'Dim OldValue = DirectCast(E.OldValue, Orientation)
            Dim NewValue = DirectCast(E.NewValue, Orientation)

            Self.OrientationCache = NewValue

            Self.InvalidateMeasure()
        End Sub

        Public Property Orientation As Orientation
            Get
                Return DirectCast(Me.GetValue(OrientationProperty), Orientation)
            End Get
            Set(ByVal value As Orientation)
                Me.SetValue(OrientationProperty, value)
            End Set
        End Property
#End Region

        Private OrientationCache As Orientation = Orientation.Vertical

    End Class

End Namespace

File: WPF/Controls/EqualizingStackPanel.cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace Ks.Common.Controls
{
    public class EqualizingStackPanel : Panel
    {

        // As Orientation.Horizontal is 0 (which makes no changes to the Orientation when got XOred with it), then the default for writing the Measure and Arrange methods will be the horizontal orientation.

        protected override Size MeasureOverride(Size AvailableSize)
        {
            var ChildCount = this.Children.Count;

            if (ChildCount == 0)
            {
                return Size.Empty;
            }

            var MaxHeight = 0.0;
            var MaxWidth = 0.0;
            AvailableSize = this.CreateSize(this.Dimension(AvailableSize, Orientation.Horizontal) / ChildCount,
                                            this.Dimension(AvailableSize, Orientation.Vertical));

            foreach (UIElement C in this.Children)
            {
                C.Measure(AvailableSize);
                var Sz = C.DesiredSize;
                MaxWidth = Math.Max(MaxWidth, this.Dimension(Sz, Orientation.Horizontal));
                MaxHeight = Math.Max(MaxHeight, this.Dimension(Sz, Orientation.Vertical));
            }

            return this.CreateSize(MaxWidth * ChildCount, MaxHeight);
        }

        protected override Size ArrangeOverride(Size FinalSize)
        {
            var ChildCount = this.Children.Count;
            var OrigFinalSize = FinalSize;

            var Width = this.Dimension(FinalSize, Orientation.Horizontal) / ChildCount;
            FinalSize = this.CreateSize(Width, this.Dimension(FinalSize, Orientation.Vertical));
            var X = 0.0;

            foreach (UIElement C in this.Children)
            {
                C.Arrange(new Rect(this.CreatePoint(X, 0), FinalSize));
                X += Width;
            }

            return OrigFinalSize;
        }

        private Size CreateSize(double Width, double Height)
        {
            if (this.OrientationCache == Orientation.Horizontal)
            {
                return new Size(Width, Height);
            }
            else
            {
                return new Size(Height, Width);
            }
        }

        private Point CreatePoint(double X, double Y)
        {
            if (this.OrientationCache == Orientation.Horizontal)
            {
                return new Point(X, Y);
            }
            else
            {
                return new Point(Y, X);
            }
        }

        private double Dimension(Size Size, Orientation Orientation)
        {
            Orientation ^= this.OrientationCache;
            if (Orientation == Orientation.Horizontal)
            {
                return Size.Width;
            }
            else
            {
                return Size.Height;
            }
        }

        public static readonly DependencyProperty OrientationProperty = DependencyProperty.Register("Orientation", typeof(Orientation), typeof(EqualizingStackPanel), new PropertyMetadata(Orientation.Vertical, Orientation_Changed, Orientation_Coerce));

        private static object Orientation_Coerce(DependencyObject D, object BaseValue)
        {
            // Dim Self = DirectCast(D, EqualizingStackPanel)

            // Dim Value = DirectCast(BaseValue, Orientation)

            return BaseValue;
        }

        private static void Orientation_Changed(DependencyObject D, DependencyPropertyChangedEventArgs E)
        {
            var Self = (EqualizingStackPanel) D;

            // Dim OldValue = DirectCast(E.OldValue, Orientation)
            var NewValue = (Orientation) E.NewValue;

            Self.OrientationCache = NewValue;

            Self.InvalidateMeasure();
        }

        public Orientation Orientation
        {
            get => (Orientation) this.GetValue(OrientationProperty);
            set => this.SetValue(OrientationProperty, value);
        }

        private Orientation OrientationCache = Orientation.Vertical;
    }
}
